common/src/main/java/com/revenuecat/purchases/common/networking/ETagManager.kt

```diff
@@ -0,0 +1,112 @@
+package com.revenuecat.purchases.common.networking
+
+import android.content.Context
+import android.content.SharedPreferences
+import com.revenuecat.purchases.common.sha1
+import org.json.JSONObject
+
+private const val NOT_MODIFIED_RESPONSE_CODE = 304
+private const val INTERNAL_SERVER_ERROR_RESPONSE_CODE = 500
+private typealias RequestHash = String
+
+data class HTTPResultWithETag(
+    val eTag: String,
+    val httpResult: HTTPResult
+) {
+    fun serialize(): String {
+        val jsonObject = JSONObject()
+        jsonObject.put("eTag", eTag)
+        jsonObject.put("httpResult", httpResult.serialize())
+        return jsonObject.toString()
+    }
+
+    companion object {
+        fun deserialize(serialized: String): HTTPResultWithETag {
+            val jsonObject = JSONObject(serialized)
+            val eTag = jsonObject.getString("eTag")
+            val serializedHTTPResult = jsonObject.getString("httpResult")
+            return HTTPResultWithETag(eTag, HTTPResult.deserialize(serializedHTTPResult))
+        }
+    }
+}
+
+class ETagManager(
+    private val prefs: SharedPreferences
+) {
+    private var hashes: Map<HTTPRequest, RequestHash> = emptyMap()
+
+    internal fun addETagHeaderToRequest(
+        httpRequestWithoutETagHeader: HTTPRequest
+    ): HTTPRequest {
+        val eTagHeader = "X-RevenueCat-ETag" to getETag(httpRequestWithoutETagHeader)
+        val updatedHeaders = httpRequestWithoutETagHeader.headers + mapOf(eTagHeader)
+        return HTTPRequest(httpRequestWithoutETagHeader.fullURL, updatedHeaders, httpRequestWithoutETagHeader.body)
+    }
+
+    internal fun processResponse(
+        httpRequest: HTTPRequest,
+        eTagInResponse: String?,
+        result: HTTPResult
+    ): HTTPResult {
+        if (eTagInResponse != null) {
+            if (result.responseCode == NOT_MODIFIED_RESPONSE_CODE) {
+                getStoredResult(httpRequest)?.let { (_, cachedResult) ->
+                    return cachedResult
+                }
```

Comment:
Oh I was returning the 305, but you're right, it would be better to reissue the request. Let's see how I can do that without going too crazy with the code